File: client/src/features/guardian/services/GuardianService.js
// @ts-nocheck
// features/guardian/api/GuardianService.js

import Api from "../../../api/Api";

class GuardianService {
    base = "/guardian";

    // ── Individual (Ward) methods ─────────────────────────────────────────────

    // Generate PIN for guardian linking
    generatePin(callbacks) {
        return Api.handleRequest(
            Api.post(`${this.base}/generate-pin`),
            callbacks
        );
    }

    // Get my current PIN
    getMyPin(callbacks) {
        return Api.handleRequest(
            Api.get(`${this.base}/my-pin`),
            callbacks
        );
    }

    // Search users to add as guardian
    searchGuardians(query, callbacks) {
        return Api.handleRequest(
            Api.get(`${this.base}/search?q=${encodeURIComponent(query)}`),
            callbacks
        );
    }

    // Get my current guardian(s)
    getMyGuardian(callbacks) {
        return Api.handleRequest(
            Api.get(`${this.base}/my-guardian`),
            callbacks
        );
    }

    // Remove my guardian
    async removeGuardian(callbacks) {
        try {
            const res = await Api.delete(`${this.base}/remove`);
            if (res.success && callbacks?.onSuccess) {
                callbacks.onSuccess(res);
            } else if (!res.success && callbacks?.onFailed) {
                callbacks.onFailed(res);
            }
        } catch (error) {
            if (callbacks?.onFailed) callbacks.onFailed(error);
        }
    }

    // ── Guardian methods ──────────────────────────────────────────────────────

    // Verify PIN and preview ward details (before linking)
    verifyPin(data, callbacks) {
        return Api.handleRequest(
            Api.post(`${this.base}/verify-pin`, data),
            callbacks
        );
    }

    // Link guardian to ward using PIN
    linkWithPin(data, callbacks) {
        return Api.handleRequest(
            Api.post(`${this.base}/link-with-pin`, data),
            callbacks
        );
    }

    // Search users to add as ward
    searchWards(query, callbacks) {
        return Api.handleRequest(
            Api.get(`${this.base}/guardian-search-wards?q=${encodeURIComponent(query)}`),
            callbacks
        );
    }

    // Get all my wards (individuals I am guardian for)
    getMyWards(callbacks) {
        return Api.handleRequest(
            Api.get(`${this.base}/guardian-my-wards`),
            callbacks
        );
    }

    // Add a ward (individual I will manage)
    addWard(data, callbacks) {
        return Api.handleRequest(
            Api.post(`${this.base}/guardian-add-ward`, data),
            callbacks
        );
    }

    // Remove a ward
    removeWard(data, callbacks) {
        return Api.handleRequest(
            Api.post(`${this.base}/guardian-remove-ward`, data),
            callbacks
        );
    }

    // ── Interest approval methods ─────────────────────────────────────────────

    // Get all pending interests for my wards
    getPendingInterests(callbacks) {
        return Api.handleRequest(
            Api.get(`${this.base}/guardian-pending-interests`),
            callbacks
        );
    }

    // Approve an interest on behalf of ward
    approveInterest(interestId, callbacks) {
        return Api.handleRequest(
            Api.post(`${this.base}/guardian-approve-interest`, { interestId }),
            callbacks
        );
    }

    // Reject an interest on behalf of ward
    rejectInterest(interestId, callbacks) {
        return Api.handleRequest(
            Api.post(`${this.base}/guardian-reject-interest`, { interestId }),
            callbacks
        );
    }

    // Get guardian pending count
    getGuardianPendingCount(callbacks) {
        return Api.handleRequest(
            Api.get(`${this.base}/guardian-pending-count`),
            callbacks
        );
    }
}

export default new GuardianService();

File: app/lib/features/guardian/services/guardian_service.dart
import 'package:get/get.dart';
import '../../../data/providers/api_client.dart';

class GuardianService extends GetxService {
  final ApiClient _api = Get.find<ApiClient>();

  // ── Ward (Individual) methods ─────────────────────────────────────────────
  Future<Map<String, dynamic>?> generatePin() async =>
      await _api.post('/guardian/generate-pin', data: {});

  Future<Map<String, dynamic>?> getMyPin() async =>
      await _api.get('/guardian/my-pin');

  Future<Map<String, dynamic>?> getMyGuardian() async =>
      await _api.get('/guardian/my-guardian');

  Future<Map<String, dynamic>?> removeGuardian() async =>
      await _api.delete('/guardian/remove');

  // ── Guardian methods ──────────────────────────────────────────────────────
  Future<Map<String, dynamic>?> verifyPin(String pin) async =>
      await _api.post('/guardian/verify-pin', data: {'pin': pin});

  Future<Map<String, dynamic>?> linkWithPin(
          String pin, String relationship) async =>
      await _api.post('/guardian/link-with-pin',
          data: {'pin': pin, 'relationship': relationship});

  Future<Map<String, dynamic>?> getMyWards() async =>
      await _api.get('/guardian/guardian-my-wards');

  Future<Map<String, dynamic>?> removeWard(dynamic individualId) async =>
      await _api.post('/guardian/guardian-remove-ward',
          data: {'wardId': individualId});

  // ── Interest approval ─────────────────────────────────────────────────────
  Future<Map<String, dynamic>?> getPendingInterests() async =>
      await _api.get('/guardian/guardian-pending-interests');

  Future<Map<String, dynamic>?> approveInterest(dynamic interestId) async =>
      await _api.post('/guardian/guardian-approve-interest',
          data: {'interestId': interestId});

  Future<Map<String, dynamic>?> rejectInterest(dynamic interestId) async =>
      await _api.post('/guardian/guardian-reject-interest',
          data: {'interestId': interestId});

  Future<Map<String, dynamic>?> getGuardianPendingCount() async =>
      await _api.get('/guardian/guardian-pending-count');
}
